[BUG] KNN Queries Fails when segment replication is enabled and replicas has deleted docs

**What is the bug?**
When segment replication is enabled on the index with more than 1 replica, exception happens when we try to do search on the replica if index contains deleted docs.

**How can one reproduce the bug?**
Steps to reproduce the behavior:
1. Setup a cluster with more than 1 node.
2. Create a knn index with segment replication enabled and atleast 1 replica.
3. Index some data in the index.
4. Delete few docs from the index.
5. Now run the search with `preference=_replica`
6. Error will be thrown.

```
PUT table
{
  "settings": {
    "index": {
      "knn": true,
      "number_of_replicas": 1
    }
  }, 
  "mappings": {
    "_source": {
      "enabled": true
    }, 
    "properties": {
      "my_vector1": {
        "type": "knn_vector",
                "dimension": 1,
                "method": {
                    "name": "hnsw",
                    "space_type": "l2",
                    "engine": "faiss"
                }
      },
      "rowId": {
        "type": "keyword",
        "store": true
      },
      "applicationId": {
        "type": "keyword",
        "store": true
      }
    }
  }
}
```
Now index data in the index.
```
PUT /_bulk?refresh=true
{ "index": { "_index": "table" } }
{"my_vector1":[1], "rowId": "recWtYS3a5LENkhte", "applicationId": "appw3EJpjHiON9qhd", "id": "1"}
{ "index": { "_index": "table"} }
{"my_vector1":[2], "rowId": "recZRfK7DnN0K0ZKM", "applicationId": "appw3EJpjHiON9qhd", "id": "2"}
{ "index": { "_index": "table"} }
{"my_vector1":[3], "rowId": "ABCrecsCBwqA7E7OWvqw", "applicationId": "appw3EJpjHiON9qhd", "id": "3"}
{ "index": { "_index": "table"} }
{"my_vector1":[4], "rowId": "recWtYS3a5LENkhte", "applicationId": "appw5EJpjHiON9qhd", "id": "4"}
{ "index": { "_index": "table"} }
{"my_vector1":[5], "rowId": "recZRfK7DnN0K0ZKM", "applicationId": "appw6EJpjHiON9qhd", "id": "5"}
{ "index": { "_index": "table"} }
{"my_vector1":[6], "rowId": "recsCBwqA7E7OWvqw", "applicationId": "appw3EJpjHiON9qhd", "id": "6"}
{ "index": { "_index": "table" } }
{"my_vector1":[7], "rowId": "recWtYS3a5LENkhte", "applicationId": "appw3EJpjHiON9qhd", "id": "7"}
{ "index": { "_index": "table"} }
{"my_vector1":[8], "rowId": "recZRfK7DnN0K0ZKM", "applicationId": "appw3EJpjHiON9qhd", "id": "8"}
{ "index": { "_index": "table"} }
{"my_vector1":[9], "rowId": "ABCrecsCBwqA7E7OWvqw", "applicationId": "appw3EJpjHiON9qhd", "id": "9"}
{ "index": { "_index": "table"} }
{"my_vector1":[10], "rowId": "recWtYS3a5LENkhte", "applicationId": "appw5EJpjHiON9qhd", "id": "10"}
{ "index": { "_index": "table"} }
{"my_vector1":[11], "rowId": "recZRfK7DnN0K0ZKM", "applicationId": "appw6EJpjHiON9qhd", "id": "11"}
{ "index": { "_index": "table"} }
{"my_vector1":[12], "rowId": "recsCBwqA7E7OWvqw", "applicationId": "appw3EJpjHiON9qhd", "id": "12"}
```
```
Delete some documents from the index.
DELETE table/_doc/<Id>
```
Now run the search by using preference as _replica
```
POST table/_search?preference=_replica
{
  "query": {
    "knn": {
      "my_vector1": {
        "vector": [
          4
        ],
        "k": 10
      }
    }
  }
}
```

Error Logs:
```
Caused by: NotSerializableExceptionWrapper[class_cast_exception: class org.apache.lucene.index.SoftDeletesDirectoryReaderWrapper$SoftDeletesFilterCodecReader cannot be cast to class org.apache.lucene.index.SegmentReader (org.apache.lucene.index.SoftDeletesDirectoryReaderWrapper$SoftDeletesFilterCodecReader and org.apache.lucene.index.SegmentReader are in unnamed module of loader 'app')]
    at org.opensearch.knn.index.query.KNNWeight.doANNSearch(KNNWeight.java:198)
    at org.opensearch.knn.index.query.KNNWeight.scorer(KNNWeight.java:122)
    at org.apache.lucene.search.Weight.scorerSupplier(Weight.java:135)
    at org.apache.lucene.search.BooleanWeight.scorerSupplier(BooleanWeight.java:596)
    at org.apache.lucene.search.Weight.bulkScorer(Weight.java:165)
    at org.apache.lucene.search.BooleanWeight.bulkScorer(BooleanWeight.java:449)
    at org.opensearch.search.internal.ContextIndexSearcher$1.bulkScorer(ContextIndexSearcher.java:383)
    at org.opensearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:324)
    at org.opensearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:283)
    at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:552)
    at org.opensearch.search.query.QueryPhase.searchWithCollector(QueryPhase.java:362)
    at org.opensearch.search.query.QueryPhase$DefaultQueryPhaseSearcher.searchWithCollector(QueryPhase.java:449)
    at org.opensearch.search.query.QueryPhase$DefaultQueryPhaseSearcher.searchWith(QueryPhase.java:433)
    at org.opensearch.search.query.QueryPhaseSearcherWrapper.searchWith(QueryPhaseSearcherWrapper.java:60)
    at org.opensearch.neuralsearch.search.query.HybridQueryPhaseSearcher.searchWith(HybridQueryPhaseSearcher.java:49)
    at org.opensearch.search.query.QueryPhase.executeInternal(QueryPhase.java:284)
    at org.opensearch.search.query.QueryPhase.execute(QueryPhase.java:157)
    at org.opensearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:556)
    at org.opensearch.search.SearchService.executeQueryPhase(SearchService.java:620)
    at org.opensearch.search.SearchService$2.lambda$onResponse$0(SearchService.java:589)
    at org.opensearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:74)
    at org.opensearch.action.ActionRunnable$2.doRun(ActionRunnable.java:89)
    at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
    at org.opensearch.threadpool.TaskAwareRunnable.doRun(TaskAwareRunnable.java:78)
    at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
    at org.opensearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:59)
    at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:922)
    at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:840)
```









**What is the expected behavior?**
Search should be successful.

**What is your host/environment?**
 - Plugins: KNN

**Do you have any screenshots?**
NA

**Do you have any additional context?**
Add any other context about the problem.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,6 +20,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Enhancements
 ### Bug Fixes
 * Fixing the arithmetic to find the number of vectors to stream from java to jni layer.[#1804](https://github.com/opensearch-project/k-NN/pull/1804)
+* Fixed LeafReaders casting errors to SegmentReaders when segment replication is enabled during search.[#1808](https://github.com/opensearch-project/k-NN/pull/1808)
 * Release memory properly for an array type [#1820](https://github.com/opensearch-project/k-NN/pull/1820)
 * FIX Same Suffix Cause Recall Drop to zero [#1802](https://github.com/opensearch-project/k-NN/pull/1802)
 ### Infrastructure

--- a/src/main/java/org/opensearch/knn/index/KNNIndexShard.java
+++ b/src/main/java/org/opensearch/knn/index/KNNIndexShard.java
@@ -10,12 +10,12 @@ import lombok.AllArgsConstructor;
 import lombok.Getter;
 import lombok.extern.log4j.Log4j2;
 import org.apache.lucene.index.FieldInfo;
-import org.apache.lucene.index.FilterLeafReader;
 import org.apache.lucene.index.IndexReader;
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.SegmentReader;
 import org.apache.lucene.store.FSDirectory;
 import org.apache.lucene.store.FilterDirectory;
+import org.opensearch.common.lucene.Lucene;
 import org.opensearch.index.engine.Engine;
 import org.opensearch.index.shard.IndexShard;
 import org.opensearch.knn.index.mapper.KNNVectorFieldMapper;
@@ -160,7 +160,7 @@ public class KNNIndexShard {
         List<EngineFileContext> engineFiles = new ArrayList<>();
 
         for (LeafReaderContext leafReaderContext : indexReader.leaves()) {
-            SegmentReader reader = (SegmentReader) FilterLeafReader.unwrap(leafReaderContext.reader());
+            SegmentReader reader = Lucene.segmentReader(leafReaderContext.reader());
             Path shardPath = ((FSDirectory) FilterDirectory.unwrap(reader.directory())).getDirectory();
             String fileExtension = reader.getSegmentInfo().info.getUseCompoundFile()
                 ? knnEngine.getCompoundExtension()

--- a/src/main/java/org/opensearch/knn/index/query/KNNWeight.java
+++ b/src/main/java/org/opensearch/knn/index/query/KNNWeight.java
@@ -11,7 +11,6 @@ import org.apache.commons.lang.StringUtils;
 import org.apache.lucene.index.BinaryDocValues;
 import org.apache.lucene.index.DocValues;
 import org.apache.lucene.index.FieldInfo;
-import org.apache.lucene.index.FilterLeafReader;
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.SegmentReader;
 import org.apache.lucene.search.DocIdSetIterator;
@@ -29,6 +28,7 @@ import org.apache.lucene.util.Bits;
 import org.apache.lucene.util.DocIdSetBuilder;
 import org.apache.lucene.util.FixedBitSet;
 import org.opensearch.common.io.PathUtils;
+import org.opensearch.common.lucene.Lucene;
 import org.opensearch.knn.common.KNNConstants;
 import org.opensearch.knn.index.KNNSettings;
 import org.opensearch.knn.index.SpaceType;
@@ -203,7 +203,7 @@ public class KNNWeight extends Weight {
 
     private Map<Integer, Float> doANNSearch(final LeafReaderContext context, final BitSet filterIdsBitSet, final int cardinality)
         throws IOException {
-        SegmentReader reader = (SegmentReader) FilterLeafReader.unwrap(context.reader());
+        final SegmentReader reader = Lucene.segmentReader(context.reader());
         String directory = ((FSDirectory) FilterDirectory.unwrap(reader.directory())).getDirectory().toString();
 
         FieldInfo fieldInfo = reader.getFieldInfos().fieldInfo(knnQuery.getField());
@@ -394,7 +394,7 @@ public class KNNWeight extends Weight {
     }
 
     private KNNIterator getFilteredKNNIterator(final LeafReaderContext leafReaderContext, final BitSet filterIdsBitSet) throws IOException {
-        final SegmentReader reader = (SegmentReader) FilterLeafReader.unwrap(leafReaderContext.reader());
+        final SegmentReader reader = Lucene.segmentReader(leafReaderContext.reader());
         final FieldInfo fieldInfo = reader.getFieldInfos().fieldInfo(knnQuery.getField());
         final BinaryDocValues values = DocValues.getBinary(leafReaderContext.reader(), fieldInfo.getName());
         final SpaceType spaceType = getSpaceType(fieldInfo);

--- a/src/test/java/org/opensearch/knn/index/SegmentReplicationIT.java
+++ b/src/test/java/org/opensearch/knn/index/SegmentReplicationIT.java
@@ -1,0 +1,94 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index;
+
+import lombok.SneakyThrows;
+import lombok.extern.log4j.Log4j2;
+import org.apache.hc.core5.http.io.entity.EntityUtils;
+import org.junit.Assert;
+import org.opensearch.client.Response;
+import org.opensearch.common.xcontent.XContentFactory;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.knn.KNNRestTestCase;
+import org.opensearch.knn.KNNResult;
+
+import java.util.List;
+
+/**
+ * This IT class contains will contain special cases of IT for segment replication behavior.
+ * All the index created in this test will have replication type SEGMENT, number of replicas: 1 and should be run on
+ * at-least 2 node configuration.
+ */
+@Log4j2
+public class SegmentReplicationIT extends KNNRestTestCase {
+    private static final String INDEX_NAME = "segment-replicated-knn-index";
+
+    @SneakyThrows
+    public void testSearchOnReplicas_whenIndexHasDeletedDocs_thenSuccess() {
+        createKnnIndex(INDEX_NAME, getKNNSegmentReplicatedIndexSettings(), createKNNIndexMethodFieldMapping(FIELD_NAME, 2));
+
+        Float[] vector = { 1.3f, 2.2f };
+        int docsInIndex = 10;
+
+        for (int i = 0; i < docsInIndex; i++) {
+            addKnnDoc(INDEX_NAME, Integer.toString(i), FIELD_NAME, vector);
+        }
+        refreshIndex(INDEX_NAME);
+        int deleteDocs = 5;
+        for (int i = 0; i < deleteDocs; i++) {
+            deleteKnnDoc(INDEX_NAME, Integer.toString(i));
+        }
+        refreshIndex(INDEX_NAME);
+        // sleep for 5sec to ensure data is replicated. I don't have a better way here to know if segments has been
+        // replicated.
+        Thread.sleep(5000);
+        // validate warmup is successful or not.
+        doKnnWarmup(List.of(INDEX_NAME));
+
+        XContentBuilder queryBuilder = XContentFactory.jsonBuilder().startObject().startObject("query");
+        queryBuilder.startObject("knn");
+        queryBuilder.startObject(FIELD_NAME);
+        queryBuilder.field("vector", vector);
+        queryBuilder.field("k", docsInIndex);
+        queryBuilder.endObject().endObject().endObject().endObject();
+
+        // validate primaries are working
+        Response searchResponse = performSearch(INDEX_NAME, queryBuilder.toString(), "preference=_primary");
+        String responseBody = EntityUtils.toString(searchResponse.getEntity());
+        List<KNNResult> knnResults = parseSearchResponse(responseBody, FIELD_NAME);
+        assertEquals(docsInIndex - deleteDocs, knnResults.size());
+
+        if (ensureMinDataNodesCountForTestingQueriesOnReplica()) {
+            // validate replicas are working
+            searchResponse = performSearch(INDEX_NAME, queryBuilder.toString(), "preference=_replica");
+            responseBody = EntityUtils.toString(searchResponse.getEntity());
+            knnResults = parseSearchResponse(responseBody, FIELD_NAME);
+            assertEquals(docsInIndex - deleteDocs, knnResults.size());
+        }
+    }
+
+    private boolean ensureMinDataNodesCountForTestingQueriesOnReplica() {
+        int dataNodeCount = getDataNodeCount();
+        if (dataNodeCount <= 1) {
+            log.warn(
+                "Not running segment replication tests named: "
+                    + "testSearchOnReplicas_whenIndexHasDeletedDocs_thenSuccess, as data nodes count is not atleast 2. "
+                    + "Actual datanode count : {}",
+                dataNodeCount
+            );
+            Assert.assertTrue(true);
+            // making the test successful because we don't want to break already running tests.
+            return false;
+        }
+        return true;
+    }
+}

--- a/src/testFixtures/java/org/opensearch/knn/KNNRestTestCase.java
+++ b/src/testFixtures/java/org/opensearch/knn/KNNRestTestCase.java
@@ -235,7 +235,11 @@ public class KNNRestTestCase extends ODFERestTestCase {
     }
 
     protected Response performSearch(final String indexName, final String query) throws IOException {
-        Request request = new Request("POST", "/" + indexName + "/_search");
+        return performSearch(indexName, query, "");
+    }
+
+    protected Response performSearch(final String indexName, final String query, final String urlParameters) throws IOException {
+        Request request = new Request("POST", "/" + indexName + "/_search?" + urlParameters);
         request.setJsonEntity(query);
 
         Response response = client().performRequest(request);
@@ -665,6 +669,35 @@ public class KNNRestTestCase extends ODFERestTestCase {
      */
     protected Settings getKNNDefaultIndexSettings() {
         return Settings.builder().put("number_of_shards", 1).put("number_of_replicas", 0).put("index.knn", true).build();
+    }
+
+    protected Settings getKNNSegmentReplicatedIndexSettings() {
+        return Settings.builder()
+            .put("number_of_shards", 1)
+            .put("number_of_replicas", 1)
+            .put("index.knn", true)
+            .put("index.replication.type", "SEGMENT")
+            .build();
+    }
+
+    @SneakyThrows
+    protected int getDataNodeCount() {
+        Request request = new Request("GET", "_nodes/stats?filter_path=nodes.*.roles");
+
+        Response response = client().performRequest(request);
+        assertEquals(RestStatus.OK, RestStatus.fromCode(response.getStatusLine().getStatusCode()));
+        String responseBody = EntityUtils.toString(response.getEntity());
+
+        Map<String, Object> responseMap = createParser(MediaTypeRegistry.getDefaultMediaType().xContent(), responseBody).map();
+        Map<String, Object> nodesInfo = (Map<String, Object>) responseMap.get("nodes");
+        int dataNodeCount = 0;
+        for (String key : nodesInfo.keySet()) {
+            Map<String, List<String>> nodeRoles = (Map<String, List<String>>) nodesInfo.get(key);
+            if (nodeRoles.get("roles").contains("data")) {
+                dataNodeCount++;
+            }
+        }
+        return dataNodeCount;
     }
 
     /**

